File: app/src/main/java/com/numero/sojodia/Fragment/BusScheduleFragment.java
package com.numero.sojodia.Fragment;

import android.os.AsyncTask;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.content.ContextCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.TextView;

import com.numero.sojodia.Manager.BusScheduleClockManager;
import com.numero.sojodia.View.TimeTableDialog;
import com.numero.sojodia.R;
import com.numero.sojodia.Util.Constants;

public class BusScheduleFragment extends Fragment implements BusScheduleClockManager.CallbackOnProgress {

    private TextView tkBusTimeTextViews[] = new TextView[3];
    private TextView tndBusTimeTextViews[] = new TextView[3];
    private TextView dateTextView;
    private BusScheduleClockManager busScheduleClockManager;
    private int reciprocating;

    public BusScheduleFragment() {
    }

    public static BusScheduleFragment newInstance(int reciprocating) {
        BusScheduleFragment fragment = new BusScheduleFragment();
        Bundle args = new Bundle();
        args.putInt(Constants.RECIPROCATING, reciprocating);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.bus_schedule_fragment, null);

        reciprocating = getArguments().getInt(Constants.RECIPROCATING);
        dateTextView = (TextView) view.findViewById(R.id.date);

        initTimeTextViews(view);
        initButtonClickListener(view);

        return view;
    }


    @Override
    public void onStart() {
        super.onStart();
        busScheduleClockManager = new BusScheduleClockManager(getContext(), reciprocating);
        busScheduleClockManager.setCallbackOnProgress(this);
        busScheduleClockManager.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
    }

    @Override
    public void onStop() {
        super.onStop();
        busScheduleClockManager.cancel(true);
    }

    private void initTimeTextViews(View view) {
        int tkTextViewIDs[] = {R.id.text_bus_time_tk, R.id.text_next_bus_time_tk, R.id.text_limit_time_tk};
        int tndTextViewIDs[] = {R.id.text_bus_time_tnd, R.id.text_next_bus_time_tnd, R.id.text_limit_time_tnd};
        for (int i = 0; i < tkTextViewIDs.length; i++) {
            tkBusTimeTextViews[i] = (TextView) view.findViewById(tkTextViewIDs[i]);
        }
        for (int i = 0; i < tndTextViewIDs.length; i++) {
            tndBusTimeTextViews[i] = (TextView) view.findViewById(tndTextViewIDs[i]);
        }
    }

    private void initButtonClickListener(View view) {
        ImageButton tkTimeTableButton = (ImageButton) view.findViewById(R.id.time_table_tk_button);
        ImageButton tndTimeTableButton = (ImageButton) view.findViewById(R.id.time_table_tnd_button);
        tkTimeTableButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                TimeTableDialog dialog = new TimeTableDialog(getActivity(), Constants.ROUTE_TK, reciprocating);
                dialog.show();
            }
        });
        tndTimeTableButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                TimeTableDialog dialog = new TimeTableDialog(getActivity(), Constants.ROUTE_TND, reciprocating);
                dialog.show();
            }
        });
    }

    @Override
    public void onProgressTK(String nextBusTimeString, String nextNextBusTimeString, String limitTimeString, int resColor) {
        tkBusTimeTextViews[0].setText(nextBusTimeString);
        tkBusTimeTextViews[1].setText(nextNextBusTimeString);
        tkBusTimeTextViews[2].setText(limitTimeString);
        tkBusTimeTextViews[2].setTextColor(ContextCompat.getColor(getContext(), resColor));
    }

    @Override
    public void onProgressTND(String nextBusTimeString, String nextNextBusTimeString, String limitTimeString, int resColor) {
        tndBusTimeTextViews[0].setText(nextBusTimeString);
        tndBusTimeTextViews[1].setText(nextNextBusTimeString);
        tndBusTimeTextViews[2].setText(limitTimeString);
        tndBusTimeTextViews[2].setTextColor(ContextCompat.getColor(getContext(), resColor));
    }

    @Override
    public void onHandlerDateChanged(String todayDateString) {
        dateTextView.setText(todayDateString);
    }
}

File: app/src/main/java/com/numero/sojodia/fragment/BusScheduleFragment.kt
package com.numero.sojodia.fragment

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.numero.sojodia.contract.BusScheduleContract
import com.numero.sojodia.model.Reciprocate
import com.numero.sojodia.model.Route
import com.numero.sojodia.util.BroadCastUtil
import com.numero.sojodia.view.adapter.BusTimePagerAdapter
import com.numero.sojodia.model.BusTime
import com.numero.sojodia.model.Time
import com.numero.sojodia.util.DateUtil
import com.numero.sojodia.util.TimeUtil
import com.numero.sojodia.view.CountDownClockTextView
import com.numero.sojodia.R
import kotlinx.android.synthetic.main.bus_schedule_fragment.*

class BusScheduleFragment : Fragment(), BusScheduleContract.View {

    private var listener: BusScheduleFragmentListener? = null

    private lateinit var presenter: BusScheduleContract.Presenter

    private var currentDateString: String? = null

    private val isDateChanged: Boolean
        get() = DateUtil.getTodayStringOnlyFigure() != currentDateString

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        if (context is BusScheduleFragmentListener) {
            listener = context
        }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        val arguments = arguments ?: return
        val reciprocate = arguments.getSerializable(Reciprocate::class.java.simpleName) as Reciprocate
        listener?.onActivityCreated(this, reciprocate)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        retainInstance = true
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        currentDateString = DateUtil.getTodayStringOnlyFigure()
        return inflater.inflate(R.layout.bus_schedule_fragment, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initCountDownClockTextView()
        initNextPreviewButton()
        initTimeTableButton()
    }

    override fun onResume() {
        super.onResume()
        presenter.subscribe()
        presenter.onTimeChanged(DateUtil.getWeek())
    }

    override fun onPause() {
        super.onPause()
        presenter.unSubscribe()
    }

    override fun setPresenter(presenter: BusScheduleContract.Presenter) {
        this.presenter = presenter
    }

    private fun initCountDownClockTextView() {
        tkCountdownTextView.setOnTimeChangedListener(object : CountDownClockTextView.OnTimeChangedListener {
            override fun onTimeChanged() {
                checkDateChange()
            }

            override fun onTimeLimit() {
                presenter.nextTkBus()
            }
        })

        tndCountdownTextView.setOnTimeChangedListener(object : CountDownClockTextView.OnTimeChangedListener {
            override fun onTimeChanged() {}

            override fun onTimeLimit() {
                presenter.nextTndBus()
            }
        })
    }

    private fun initTimeTableButton() {
        tkTimeTableButton.setOnClickListener {
            presenter.showTimeTableDialog(Route.TK)
        }
        tndTimeTableButton.setOnClickListener {
            presenter.showTimeTableDialog(Route.TND)
        }
    }

    private fun initNextPreviewButton() {
        tkNextImageButton.setOnClickListener {
            presenter.nextTkBus()
        }
        tkPreviewImageButton.setOnClickListener {
            presenter.previewTkBus()
        }
        tndNextImageButton.setOnClickListener {
            presenter.nextTndBus()
        }
        tndPreviewImageButton.setOnClickListener {
            presenter.previewTndBus()
        }
    }

    private fun checkDateChange() {
        if (isDateChanged) {
            currentDateString = DateUtil.getTodayStringOnlyFigure()
            BroadCastUtil.sendBroadCast(activity, BroadCastUtil.ACTION_CHANGED_DATE)
            presenter.onTimeChanged(DateUtil.getWeek())
        }
    }

    override fun showTkBusTimeList(busTimeList: MutableList<BusTime>) {
        tkBusTimeViewPager.apply {
            adapter = BusTimePagerAdapter().apply {
                setBusTimeList(busTimeList)
            }
        }
    }

    override fun showTndBusTimeList(busTimeList: MutableList<BusTime>) {
        tndBusTimeViewPager.apply {
            adapter = BusTimePagerAdapter().apply {
                setBusTimeList(busTimeList)
            }
        }
    }

    override fun selectTkCurrentBusPosition(position: Int) {
        tkBusTimeViewPager.setCurrentItem(position, true)
    }

    override fun startTkCountDown(busTime: BusTime) {
        val time = Time(busTime.hour, busTime.min, 0)
        val countTime = TimeUtil.getCountTime(time)
        tkCountdownTextView.setTime(countTime.hour, countTime.min)
    }

    override fun selectTndCurrentBusPosition(position: Int) {
        tndBusTimeViewPager.setCurrentItem(position, true)
    }

    override fun startTndCountDown(busTime: BusTime) {
        val time = Time(busTime.hour, busTime.min, 0)
        val countTime = TimeUtil.getCountTime(time)
        tndCountdownTextView.setTime(countTime.hour, countTime.min)
    }

    override fun showTkNextButton() {
        tkNextImageButton.visibility = View.VISIBLE
    }

    override fun showTkPreviewButton() {
        tkPreviewImageButton.visibility = View.VISIBLE
    }

    override fun showTkNoBusLayout() {
        tkNoBusLayout.visibility = View.VISIBLE
    }

    override fun showTndNextButton() {
        tndNextImageButton.visibility = View.VISIBLE
    }

    override fun showTndPreviewButton() {
        tndPreviewImageButton.visibility = View.VISIBLE
    }

    override fun showTndNoBusLayout() {
        tndNoBusLayout.visibility = View.VISIBLE
    }

    override fun hideTkNextButton() {
        tkNextImageButton.visibility = View.GONE
    }

    override fun hideTkPreviewButton() {
        tkPreviewImageButton.visibility = View.GONE
    }

    override fun hideTkNoBusLayout() {
        tkNoBusLayout.visibility = View.GONE
    }

    override fun hideTndNextButton() {
        tndNextImageButton.visibility = View.GONE
    }

    override fun hideTndPreviewButton() {
        tndPreviewImageButton.visibility = View.GONE
    }

    override fun hideTndNoBusLayout() {
        tndNoBusLayout.visibility = View.GONE
    }

    override fun showTimeTableDialog(route: Route, reciprocate: Reciprocate) {
        listener?.showTimeTableDialog(route, reciprocate)
    }

    interface BusScheduleFragmentListener {
        fun onActivityCreated(fragment: BusScheduleFragment, reciprocate: Reciprocate)

        fun showTimeTableDialog(route: Route, reciprocate: Reciprocate)
    }

    companion object {
        fun newInstance(reciprocate: Reciprocate): BusScheduleFragment = BusScheduleFragment().apply {
            arguments = Bundle().apply {
                putSerializable(Reciprocate::class.java.simpleName, reciprocate)
            }
        }
    }
}
